components/learn/modules/ROOT/pages/unit-testing.adoc

```diff
@@ -0,0 +1,174 @@
+
+Time to write our first tests! These will test properties of the `SetAndGet` contract xref:write-contracts#set-and-get-contract[from previous guides].
+
+```javascript
+// Load dependencies
+const { accounts, contract } = require('@openzeppelin/test-environment');
+const { expect } = require('chai');
+
+// Load compiled artifacts
+const SetAndGet = contract.fromArtifact('SetAndGet');
+
+// Start test block
+describe('SetAndGet', function () {
+  const [ owner ] = accounts;
+
+  beforeEach(async function () {
+    // Deploy a new contract for each test
+    this.contract = await SetAndGet.new({ from: owner });
+  });
+
+  // Test case
+  it('get returns a value previously set', async function () {
+    // Set a value
+    await this.contract.set(42, { from: owner });
+
+    // Test the returned value is the same one - values from the blockchain are
+    // often read as strings
+    expect((await this.contract.get()).toString()).to.equal('42');
+  });
+});
+```
+
+TIP: Many books have been written about how to structure unit tests: for a quick reference, check out the https://github.com/OpenZeppelin/openzeppelin-contracts/tree/master/test[tests for OpenZeppelin Contracts], or the following https://medium.com/coinmonks/how-to-test-ethereum-smart-contracts-ac28fa852281[guide for ERC20 tests].
+
+We are now ready to run our tests! The best way to do this is by adding a `test` script to your `package.json`:
+
+[source,diff]
+----
+ 'scripts': {
++  'test': 'mocha --exit --recursive test'
+ }
+----
+
+With this, running `npm test` will execute all tests in the `test` directory, checking that your contracts work the way you meant them too.
+
+TIP: Don't forget to re-compile your contracts if you make changes to them!
+
+It's also a very good idea at this point to set up a Continuous Integration service such as https://circleci.com/[CircleCI] to make your tests run automatically every time you commit your code to GitHub.
+
+[[performing-complex-assertions]]
+== Performing Complex Assertions
+
+Many interesting properties of your contracts can be hard to capture with JavaScript primitives, such as:
+
+ * verifying that the contract reverts on errors
+ * measuring by how much an account's Ether balance changed
+ * validating the values of emitted events
+
+https://github.com/OpenZeppelin/openzeppelin-test-helpers#readme[*OpenZeppelin Test Helpers*] is a library designed to help you test all of these properties. It will also let you simulate time passing on the blockchain, handle very large numbers, and have easy access to common constants.
+
+Begin by installing it:
+
+```bash
+$ npm install --save-dev @openzeppelin/test-helpers
+```
+
+Because Test Environment will detect this installation and configure it for you, we can get started using the library right away:
```

Comment:
> Because Test Environment will detect this installation and configure it for you

Let's push this down to after we show how to use it. Here I'd just mention to get started using it.